bpf/lib/common.h

```diff
@@ -349,16 +357,17 @@ static inline union v6addr *
 ipv6_ct_tuple_get_daddr(struct ipv6_ct_tuple *tuple)
 {
 #ifdef CONNTRACK
-	/* For outgoing connections, lib/conntrack.h swaps the src/dst. */
-	return &tuple->saddr;
-#else
 	return &tuple->daddr;
+#else
+	return &tuple->saddr;
 #endif
 }
 
+/* Tuple addresses and ports are loaded from a packet in the reverse order.
+ * For outgoing connections, lib/conntrack swaps them to the correct order. */
```

Comment:
I wonder if it's worth expanding on this a bit more (+ipv6 equivalent).

Basically we've ended up in a situation where the `ipv4_ct_tuple` never corresponds directly, byte-by-byte, to the contents of a packet we've seen.

When comparing the names of the fields and ordering, the addresses are the same order between `struct ipv4_ct_tuple` and `struct iphdr`, however we load them in the reverse order as follows (from above in this commit):

```
+       tuple.daddr = ip4->saddr;
+       tuple.saddr = ip4->daddr;
```

Then the ports fields of `struct ipv4_ct_tuple` are in the opposite order of `struct udphdr`, and we load these byte-by-byte, so in memory they're in the same order but the field names reference them in the opposite way:

https://github.com/cilium/cilium/blob/71a863422a43e0499cd9b777b68261227544de2a/bpf/lib/conntrack.h#L390-L391

This might be relevant background for any new usages of the tuple, but is unlikely to catch developers out unless they attempt to do byte-wise comparison/copying of the fields.

We could also mention that this is now used across Cilium restarts so the ordering cannot be modified.

Another comment that would be useful is the description of the field names with respect to packet contents, eg:

```
The source/destination fields in this structure indicate the source/destination addresses/ports of the original direction of the connection. Users typically populate the ipv4_ct_tuple with the reverse direction first (ie, populate 'saddr' with the packet's 'daddr', 'sport' with 'dport', etc., attempt a CT look up, then if this fails (indicating that there is no corresponding original direction connection), reverse the tuple which makes the sources/destinations match the packet's sources/destinations. At this point, the CT entry will be committed (modulo policy), which means that subsequent replies, following the same logic as above, would find this newly created entry.
```